sources/whippet/src/policy.rs

```diff
@@ -0,0 +1,897 @@
+
+        Ok(policy)
+    }
+
+    /// Sets the priorities for the rules of a policy, starting with the users' rules
+    pub(crate) fn set_rule_priorities(&mut self, current_priority: &mut u64) {
+        if let Some(user_policy) = &mut self.user {
+            for policy in user_policy.values_mut() {
+                for rule in &mut policy.rules {
+                    let priority = PRIORITY_BASE * DEFAULT_POLICY_CONTEXT_USER + *current_priority;
+                    rule.set_priority(priority);
+                    *current_priority += 1;
+                }
+            }
+        }
+        if let Some(default_rules) = &mut self.default {
+            for rule in &mut default_rules.rules {
+                let priority =
+                    PRIORITY_BASE * DEFAULT_POLICY_CONTEXT_MULTIPLIER + *current_priority;
+                rule.set_priority(priority);
+                *current_priority += 1;
+            }
+        }
+    }
+
+    /// Validates that the user contexts don't include connect rules
+    fn validate(&self) -> Result<()> {
+        if let Some(user_policy) = &self.user {
+            for context in user_policy.values() {
+                ensure!(
+                    !context
+                        .rules
+                        .iter()
+                        .any(|rule| matches!(rule, Rule::ConnectUser { .. })),
+                    error::UserContextWithConnectRuleSnafu
+                );
+            }
+        }
+        Ok(())
+    }
+
+    /// Replaces the UIDs of the users that were found in the policies
+    fn replace_uids<U>(&mut self, username_resolver: &mut U) -> Result<()>
+    where
+        U: UsernameResolver,
+    {
+        // First, replace the users of all the connect rules in the default context
+        if let Some(default_context) = &mut self.default {
+            default_context
+                .rules
+                .iter_mut()
+                .try_for_each(|rule| match rule {
+                    Rule::ConnectUser { user, .. } => {
+                        *user = username_resolver.resolve(user)?.to_string();
+                        Ok(())
+                    }
+                    _ => Ok(()),
+                })?;
+        };
+
+        // Then, replace all the users of the users' context. Connect rules are not allowed
+        // in other contexts other than the default context, so the rules don't have to be modified
+        if let Some(ref mut users_contexts) = &mut self.user {
+            let mut final_users_contexts = IndexMap::new();
+            let users: Vec<String> = users_contexts.keys().cloned().collect();
+            for user in users {
+                let uid = username_resolver.resolve(&user)?.to_string();
+                let policy = users_contexts.shift_remove(&user).unwrap_or_default();
+                final_users_contexts.insert(uid, policy);
+            }
+            self.user = Some(final_users_contexts);
+        }
```

Comment:
Try something like this to avoid all the shifting and cloning:
```suggestion
        let users_contexts = self.user.take().unwrap_or_default();
        let mut final_users_contexts = IndexMap::new();
        for (user, policy) in users_contexts.into_iter() {
            let uid = username_resolver.resolve(&user)?.to_string();
            final_users_contexts.insert(uid, policy);
        }
        self.user = Some(final_users_contexts);
```